Import form refactor with FormikWizard

rohitkrai03 opened the pull request:
## Fixes 
<!-- For e.g Fixes: https://issues.redhat.com/browse/HAC-XXX -->
- https://issues.redhat.com/browse/HAC-1604
- https://issues.redhat.com/browse/HAC-657

## Description
<!-- Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change. -->
- Redesigns the import form with formik wizard.
- Refactors all the import forms.

## Type of change
<!-- Please delete options that are not relevant. -->

- [x] Feature
- [ ] Bugfix
- [ ] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## Screen shots / Gifs for design review 
<!-- If change affects UI in any way, tag relevant UX people and add screenshots/gifs  -->


https://user-images.githubusercontent.com/6041994/179459832-8e5d328b-d646-4f76-85e7-38a37a2da2d8.mov


## Browser conformance: 
<!-- To mark tested browsers, use [x] -->
- [x] Chrome
- [ ] Firefox
- [ ] Safari
- [ ] Edge

rohitkrai03 commented:
@openshift/team-hacdev-ux Needs reviews from UX.

rottencandy commented:
When entering a private repo, the next button's spinner is always displayed, even after binding tokens are created. I think it might confuse users.
Would it be possible to only show the spinner when the next button is clicked?
![0](https://user-images.githubusercontent.com/20013884/179520640-d506bb15-734c-4dcb-992b-7844cd8bb8ab.png)

rottencandy reviewed (commented):
Some more next button-related issues:
- Next button on source page is initially enabled, in both add application & add comp forms
(although it gets disabled when clicked)
![0](https://user-images.githubusercontent.com/20013884/179524543-45b7ee7d-301e-4407-b087-217435ef46a5.png)
- Navigating back to application page using steps causes next button to sometimes be disabled

https://user-images.githubusercontent.com/20013884/179524653-b88d89cb-b973-49a5-8135-c23ff48e469a.mp4

rohitkrai03 commented:
> When entering a private repo, the next button's spinner is always displayed, even after binding tokens are created. I think it might confuse users. Would it be possible to only show the spinner when the next button is clicked? ![0](https://user-images.githubusercontent.com/20013884/179520640-d506bb15-734c-4dcb-992b-7844cd8bb8ab.png)

Fixed it. Removed the loading indicator in next button.

rohitkrai03 commented on src/components/ComponentSettingsForm/ComponentSettingsForm.tsx:
```diff
@@ -48,16 +48,15 @@ const ComponentSettingsForm: React.FunctionComponent<ComponentSettingsFormProps>
-              isExpanded
-              editMode
-            />
+            {components.map((component) => (
```
Not really but I just wanted to be consistent of how we render the review cards. We might be looking into completely changing the edit components form in future though.

rohitkrai03 commented on src/components/ImportForm/ImportForm.tsx:
```diff
@@ -0,0 +1,73 @@
+        onSubmit={handleSubmit}
+        onReset={handleReset}
+        initialValues={initialValues}
+        steps={steps as any}
```
No. I think it was from before I was still testing a few things with wizard steps.

rohitkrai03 commented on src/components/ImportForm/ReviewSection/ReviewSection.tsx:
```diff
@@ -0,0 +1,135 @@
+    values: { source, secret, application, git },
+    setFieldValue,
+  } = useFormikContext<ImportFormValues>();
+  const cachedComponents = React.useRef([]);
+  const cachedComponentsLoaded = React.useRef(false);
```
Yeah and also because we don't wanna re-render when user updates a field in the form.

rohitkrai03 commented on integration-tests/support/pages/AddComponentPage.ts:
```diff
@@ -1,16 +1,16 @@
 import { Common } from '../../utils/Common';
-import { pageTitles } from '../constants/PageTitle';
+// import { pageTitles } from '../constants/PageTitle';
```
Removed.

rottencandy commented on src/components/ImportForm/utils/validation-utils.ts:
```diff
@@ -0,0 +1,73 @@
+    .string()
+    .required('Required')
+    .test('isUnique', 'Application name already exists.', async (value, context) => {
+      try {
+        await createApplication(value, context.parent.namespace, true);
+        return true;
```
Can this be debounced? Right now it's running on every keystroke

rottencandy replied:
Okay, using `validateOnChange` in the step.
But not sure if it's working, I can see multiple network calls on every key press.

rohitkrai03 commented on src/components/ImportForm/SampleSection/SampleSection.tsx:
```diff
@@ -0,0 +1,180 @@
+                <Button
+                  variant={ButtonVariant.link}
+                  onClick={handleStrategyChange}
+                  style={{ paddingLeft: 0 }}
```
Done.

rottencandy reviewed (approved):
/lgtm
